File: vi/codes/swift/chapter_stack_and_queue/queue.swift
/**
 * File: queue.swift
 * Created Time: 2023-01-11
 * Author: nuomi1 (nuomi1@qq.com)
 */

@main
enum Queue {
    /* Mã điều khiển */
    static func main() {
        /* Truy cập phần tử ở đầu hàng đợi */
        // Swift không có lớp hàng đợi tích hợp sẵn, có thể dùng Array làm hàng đợi
        var queue: [Int] = []

        /* Thêm phần tử vào hàng đợi */
        queue.append(1)
        queue.append(3)
        queue.append(2)
        queue.append(5)
        queue.append(4)
        print("Queue queue = \(queue)")

        /* Trả về danh sách để in ra */
        let peek = queue.first!
        print("Front element peek = \(peek)")

        /* Xóa phần tử khỏi hàng đợi */
        // Khi mô phỏng bằng Array, độ phức tạp của pop là O(n)
        let pool = queue.removeFirst()
        print("Dequeue element pop = \(pool), after dequeue queue = \(queue)")

        /* Lấy độ dài của hàng đợi */
        let size = queue.count
        print("Queue length size = \(size)")

        /* Kiểm tra hàng đợi có rỗng không */
        let isEmpty = queue.isEmpty
        print("Is queue empty = \(isEmpty)")
    }
}

File: vi/codes/cpp/chapter_stack_and_queue/queue.cpp
/**
 * File: queue.cpp
 * Created Time: 2022-11-28
 * Author: qualifier1024 (2539244001@qq.com)
 */

#include "../utils/common.hpp"

/* Driver Code */
int main() {
    /* Khởi tạo hàng đợi */
    queue<int> queue;

    /* Thêm phần tử vào hàng đợi */
    queue.push(1);
    queue.push(3);
    queue.push(2);
    queue.push(5);
    queue.push(4);
    cout << "Queue queue = ";
    printQueue(queue);

    /* Truy cập phần tử ở đầu hàng đợi */
    int front = queue.front();
    cout << "Front element front = " << front << endl;

    /* Xóa phần tử khỏi hàng đợi */
    queue.pop();
    cout << "Dequeue element front = " << front << ", after dequeue, queue = ";
    printQueue(queue);

    /* Lấy độ dài của hàng đợi */
    int size = queue.size();
    cout << "Queue length size = " << size << endl;

    /* Kiểm tra hàng đợi có rỗng không */
    bool empty = queue.empty();
    cout << "Queue is empty = " << empty << endl;

    return 0;
}
